In [1]:
#this is the DS6: Data question1
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

Using the pandas read_csv() function, read the GDP dataset into your notebook as a DataFrame called gdp_df.
Take a look at the first 10 rows.
Look at the last 5 rows. Do you see a problem? 
    
Redo the read_csv() call to correct this issue - do not modify the original csv file.

In [15]:
#Reading the gdp_per_capita.csv which is downloaaded fro the URL("") as .csv file and nested in the data folder
gdp_df = pd.read_csv("../data/gdp_per_capita.csv")
gdp_df.head(10)
gdp_df.tail(5)

# there are 4 coulmns Country or Area, Year, Value, Value Footnotes. Value Footnotes has NaN values

Country or Area                                      Year        Value  \
6866        Zimbabwe                                      1991  3426.598094   
6867        Zimbabwe                                      1990  3324.348171   
6868             NaN                                       NaN          NaN   
6869   footnoteSeqID                                  Footnote          NaN   
6870               2  Excludes South Sudan after July 9, 2011.          NaN   

      Value Footnotes  
6866              NaN  
6867              NaN  
6868              NaN  
6869              NaN  
6870              NaN

In [16]:
gdp_df.shape #(6871, 4)
gdp_df.info()
gdp_df.columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6871 entries, 0 to 6870
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  6870 non-null   object 
 1   Year             6870 non-null   object 
 2   Value            6868 non-null   float64
 3   Value Footnotes  1 non-null      float64
dtypes: float64(2), object(2)
memory usage: 214.8+ KB


Index(['Country or Area', 'Year', 'Value', 'Value Footnotes'], dtype='object')

1. There are 7871 rows and 4 colums

2. There are 4 coulmns Country or Area, Year, Value, Value Footnotes.
Index(['Country or Area', 'Year', 'Value', 'Value Footnotes'], dtype='object')

2. The df has 6871 entries (rows of information)
RangeIndex: 6871 entries, 0 to 6870
Data columns (total 4 columns):
   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  6870 non-null   object 
 1   Year             6870 non-null   object 
 2   Value            6868 non-null   float64
 3   Value Footnotes  1 non-null      float64
dtypes: float64(2), object(2)

2. Value Footnotes has NaN values.
3. Yes there are NaN values in the country column and Year. The world Footnote in the Year column changes the Year column type to object, rather than keeping it to integer.



In [17]:
# droping the NaN values in the df columns country and value
gdp_df_dropnan = gdp_df.dropna(subset=['Country or Area', 'Value']) 
gdp_df_dropnan.shape
gdp_df_nan.tail(5)
gdp_df_nan.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6868 entries, 0 to 6867
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  6868 non-null   object 
 1   Year             6868 non-null   object 
 2   Value            6868 non-null   float64
 3   Value Footnotes  1 non-null      float64
dtypes: float64(2), object(2)
memory usage: 268.3+ KB


This gives us the df with 6868 rows and 4 columns.

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6868 entries, 0 to 6867
Data columns (total 4 columns):
    Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  6868 non-null   object 
 1   Year             6868 non-null   object 
 2   Value            6868 non-null   float64
 3   Value Footnotes  1 non-null      float64
dtypes: float64(2), object(2)
memory usage: 268.3+ KB

In [19]:
#Drop the 'Value Footnotes' column, and rename the remaining columns to 'Country', 'Year', and 'GDP_Per_Capita'

gdp_df_dropnan_dropfootnotes = gdp_df_dropnan.drop(columns = ["Value Footnotes"])
print(gdp_df_dropnan_dropfootnotes.tail(5))
gdp_df_dropnan_dropfootnotes.columns

     Country or Area  Year        Value
6863        Zimbabwe  1994  3263.934978
6864        Zimbabwe  1993  3033.504852
6865        Zimbabwe  1992  3054.889178
6866        Zimbabwe  1991  3426.598094
6867        Zimbabwe  1990  3324.348171


Index(['Country or Area', 'Year', 'Value'], dtype='object')

In [22]:
gdp_df_column_rename = gdp_df_dropnan_dropfootnotes.rename(columns = {'Country or Area' : 'Country', 'Year' : 'Year', 'Value' : 'GDP_Per_Capita'})
print(gdp_df_column_rename.head(5))
gdp_df_column_rename.info() # we can also use the df.dtype() to find the type.

       Country  Year  GDP_Per_Capita
0  Afghanistan  2019     2065.036398
1  Afghanistan  2018     2033.779002
2  Afghanistan  2017     2058.383832
3  Afghanistan  2016     2057.062164
4  Afghanistan  2015     2068.265904
<class 'pandas.core.frame.DataFrame'>
Int64Index: 6868 entries, 0 to 6867
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country         6868 non-null   object 
 1   Year            6868 non-null   object 
 2   GDP_Per_Capita  6868 non-null   float64
dtypes: float64(1), object(2)
memory usage: 214.6+ KB


The Nan values are taken out, footnotes removed.
Now we have 6868 rows, and 3 columns, but the type is the object for country name and year. we want the year to be the int

In [24]:
#convert the type using the .astype() function for the column, and selecting it to be assinged back sort of saving it in the same dataframe to the same df
gdp_df_column_rename["Year"] = gdp_df_column_rename["Year"].astype(int)
gdp_df_column_rename.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6868 entries, 0 to 6867
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country         6868 non-null   object 
 1   Year            6868 non-null   int32  
 2   GDP_Per_Capita  6868 non-null   float64
dtypes: float64(1), int32(1), object(1)
memory usage: 187.8+ KB


This CSV could have been read differently using the nrows function to drop the last row in the year column which has a string and converted it to an object column.

In [28]:
# Reread the .csv file into the dataframe without changing the datatype and droping the Footnotes row.
gdp_df_nrows = pd.read_csv("../data/gdp_per_capita.csv", nrows=6868)
gdp_df_nrows.info()
gdp_df_nrows.shape
gdp_df_nrows.columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6868 entries, 0 to 6867
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  6868 non-null   object 
 1   Year             6868 non-null   int64  
 2   Value            6868 non-null   float64
 3   Value Footnotes  1 non-null      float64
dtypes: float64(2), int64(1), object(1)
memory usage: 214.8+ KB


Index(['Country or Area', 'Year', 'Value', 'Value Footnotes'], dtype='object')

We now have a df called gdp_df_nrows where there are 4 columns and 6868 rows. The datatype is object for country, int for year and float for GDP and footnotes
Now will drop the Footnotes coulmn and rename the coulmn header.

In [30]:
gdp_df_for_analysis = gdp_df_nrows.drop(columns = ["Value Footnotes"])
gdp_df_for_analysis = gdp_df_for_analysis.rename(columns = {'Country or Area': 'Country', 'Year': 'Year', 'Value': 'GDP_Per_Capita'})
gdp_df_for_analysis.head(5)

Country  Year  GDP_Per_Capita
0  Afghanistan  2019     2065.036398
1  Afghanistan  2018     2033.779002
2  Afghanistan  2017     2058.383832
3  Afghanistan  2016     2057.062164
4  Afghanistan  2015     2068.265904

6. How many rows and columns does gdp_df have? What are the data types of its columns? If any of the columns are not the expected types, figure out why and fix it.

In [31]:
gdp_df_for_analysis.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6868 entries, 0 to 6867
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country         6868 non-null   object 
 1   Year            6868 non-null   int64  
 2   GDP_Per_Capita  6868 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 161.1+ KB


 How many rows and columns does gdp_df have? 
     Rows = 6868 and Columns: 3
 What are the data types of its columns? 
     Country = Object
     Year = Int
     GDP_Per_Capita = float
         
If any of the columns are not the expected types, figure out why and fix it.
    Column type and value are fine.

Which years are represented in this dataset? Take a look at the number of observations per year. What do you notice?